pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,2106 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict, namedtuple
+from mpi4py import MPI
+import numpy as np
+import os
+import re
+import sys
+
+from pyfr.inifile import Inifile
+from pyfr.mpiutil import get_comm_rank_root, get_mpi
+from pyfr.nputil import CubicSplineFit, fuzzysort, LinearFit, npeval
+from pyfr.plugins.base import BasePlugin
+from pyfr.writers.native import NativeWriter
+
+
+def open_csv(fname, header=None, mode='w'):
+    # Open for mode
+    outf = open(fname, mode)
+    # Output a header if required
+    if os.path.getsize(fname) == 0 and header:
+        print(header, file=outf)
+    return outf
+
+def write_fftdata(fname, *indata, header='', mode='a'):
+    #open the file
+    outf = open_csv(fname, ''.join(header), mode)
+    #prepare the data array
+    data = np.array(indata)
+    dshape = data.shape
+    #do write
+    for i in range(dshape[1]):
+        print(*data[:, i].ravel(), sep=' ', file=outf)
+    outf.flush()
+    return
+
+def realfft(udata_):
+        dsize = udata_.shape[-1]
+        # freqsize: size of half the spectra with positive frequencies
+        freqsize = int(dsize/2) if dsize%2 == 0 else int(dsize-1)/2
+        ufft = np.fft.rfft(udata_)/freqsize 
+        return ufft
+
+def compute_spl(pref, amp, df=None):
+    sqrt2 = 1./np.sqrt(2.) 
+    lgsqrt2 = 20.*np.log10(sqrt2)
+    factor = sqrt2/pref if not df else sqrt2/(pref*np.sqrt(df))
+    spl = 20.*np.log10(factor*amp) 
+    spl[0] -= lgsqrt2
+    spl[-1] -= lgsqrt2
+    #overall sound pressure level
+    ospl_sum = sum(pow(10., spl/10.)) if not df else df*sum(pow(10., spl/10.))
+    oaspl = 10.* np.log10(ospl_sum) 
+    return spl, oaspl
+
+def compute_psd(amp, df):
+    psd = 0.5*amp*amp/df
+    psd[0] *= 2.
+    psd[-1] *= 2.
+    return psd
+
+def Gatherv_data_arr(comm, rank, root, data_arr):
```

Comment:
This function is silly.  The purpose of the capital letter methods in mpi4py is to reduce overhead at the cost of additional complexity.  But here, we get the worse of both worlds: the inefficiency of the lower case method with the complexity of the upper case method.  Just use `comm.gather(data_arr, root=root)` unless you're going to have the root rank cache the lengths in advance.

Reply:
This function Gatherv is used in multiple places and it is more compact to have a method for it here. Also, the arrays I send and receive are variable so "v" is needed and they are numpy arrays and I prefer to work with them and be able to manipulate the shape easier.

Reply:
Read my comment again.  This entire function is just a poor reimplementaton of `comm.gather`.  Every use of it can be replaced with `comm.gather(data_arr, root=root)` and will deliver similar performance.  And the lower case `.gather` works with everything (Python objects, NumPy arrays, whatever).

Reply:
Removed most of the capital letter mpi communications. Can change to capital letter after testing if needed with the help of cached counts and displs.